### DISCUSSION TITLE
Proposal - Define JSON Logic on the JSON Data Model

### DISCUSSION THREAD

**Discussion by gregsdennis** [json-logic/.github#17] [Ideas]
JSON Schema, JSON Path, and JSON-e are all defined on the JSON data model.  That is, they are defined on a data model that consists of:

- objects (unordered string-key/value pairs)
- arrays (ordered [not sorted] values)
- strings
- numbers
- `true` literal
- `false` literal
- `null` literal

`true` and `false` are typically combined as a "boolean" type.  Also, an object property being present with a `null` value is distinct from the property missing. 

Most importantly, operating on this data model:

- removes any lexical issues from the scope of the spec (whitespace, number formatting, etc.)
- allows processing of other compatible formats, e.g. YAML

---

> [!IMPORTANT]  
> **@TotalTechGeek is appending the following suggestion to this proposal https://github.com/orgs/json-logic/discussions/17#discussioncomment-11674812**

**Comment by JTeeuwissen** [comment]
Not sure if within the scope of this proposal, but I think most programming languages can describe this model relatively straightforward, but I think numbers require some extra considerations. Should they be represented as floating binary points or floating decimal points, and how precise? Json itself does not put any constraints on the precision of numbers, but that might result in an inefficient or unworkable solution in most languages.

**Reply by TotalTechGeek** [reply]
This is worth discussing here, I think

**Reply by gregsdennis** [reply]
JSON Schema and Path retain the lack of restriction on numbers that JSON describes.  This passes the consideration to implementations. 

Specifically, it means that implementations can indicate their own limitations, whether self-imposed or by the language.

The Schema test suite does also have some optional tests to check for large values and high precision.

**Reply by rmannibucau** [reply]
Several JSON implementations - (de)serializer - use ieee754. Main issue being when in the language you have "big numbers" and primitives, if you want to be clever and serialize big numbers as double/json numbers you end up being polymorphic (string or number) which is very bad, so putting limitations there would be good IMHO.

One sample discussion (but there are tons): https://github.com/jqlang/jq/issues/143

**Reply by gregsdennis** [reply]
Yes the problem is the parsers.  I have a blog post on this topic. 

https://blog.json-everything.net/posts/numbers-are-numbers-not-strings/

That said, it's not the responsibility of the spec to put limitations on this. Moreover, different languages will have different limitations, meaning that they'll have varying levels of support  regardless of what we specify. 

The solution is to leave it entirely with implementations to state what they support.

**Reply by dslmeinte** [reply]
(As an anecdotal aside: numbers in strings being parsed *inaccurately* as floating points numbers was the pivot I needed to decide to start with CertLogic, so yes: parsing is the problem — also here.)

I think the JSON Data Model should work fine _as a basis_ for the specification, but the JSON Data Model (not even when combined with JSON Schema and the likes) probably provide all the expressive power you need for an actual specification — especially not where it concerns the semantics.

**Reply by rmannibucau** [reply]
> That said, it's not the responsibility of the spec to put limitations on this.

I think it is, typicall within JSON-P/JSON-B (java/jakarta land) some limitations had been explicitly stated on what is supported in the link between json and java, JSON-Logic should likely explicit what is supposed to be supported by an implementation and then the language support *MUST* make it work - using strings internally if needed.

That said I ack if we can cover most languages and arch it would be good so IEEE754 is a good start IMHO.

Side note: can also mean `var` replacement can be associated with a new `type` metadata to refine it later - thinking out loud.

**Reply by TotalTechGeek** [reply]
I actually think I find myself aligned with @gregsdennis here. 

I'm comfortable with the route similar projects took w/r/t JSON and the opinions expressed in his article.

_However_

I think it'd be a good idea to document & test numeric precision in our compatibility tables; I think that IEEE754 compat will _generally_ be a safe assumption. But we can communicate this in tables.

_Also_

**Outside of JSON Logic Core,** I could see a common community extension that has `type` operators for coercion, like `{ "bigint": "12345" }`. (Which I'd concur with @gregsdennis is indeed a bit of an anti-pattern / workaround, instead of asking more of our JSON Parsers)

**Reply by gregsdennis** [reply]
> typicall within JSON-P/JSON-B (java/jakarta land)

It's fine for a technology to specify these limitations if they're specific to a language because said language has those limitations.  But we're developing a cross-language, interoperable technology, where consistency cannot be guaranteed.

**Reply by rmannibucau** [reply]
> if they're specific to a language because said language has those limitations

note that language here is literally JSON, there is no uniform standard accross JSON specs (and worse when you take javascript - as an ecosystem not last V8 or node version.
parsers can't help that there is nothing defined and that machines have limitations, depending if you run on a 32 bits  or 64 bits machines you are not able to support the same precision for ex so something must be defined IMHO even if spec is written with something like "we guarantee xxxx but note that implementations are free to support more but this is not portable"

**Reply by gregsdennis** [reply]
RFC 8259 and all of the other JSON specs are consistent in defining numbers as infinite range and precision.  The latest has a note indicating that implementations can place limitations as the need to. 

https://datatracker.ietf.org/doc/html/rfc8259#section-6

This is the pattern we should follow: we define no restrictions, but implementations can.

**Reply by rmannibucau** [reply]
> This is the pattern we should follow: we define no restrictions, but implementations can.

it is the other way around if you want to do a spec: you give guarantees and open for more features but this is no more under the spec imbrella.
If we don't do that there is no real value to do a spec since nothing will be portable, not even tests since you will not be able to write it since nothing will be defined.

side note: the part you reference encourages to use ieee754 and mentions that anything else is not portable (= out of scope of the spec even if not forbidden), so guess we should do it which means using ieee754

**Reply by gregsdennis** [reply]
> it is the other way around if you want to do a spec: you give guarantees and open for more features but this is no more under the spec imbrella.

No.  Having worked on multiple specifications, and looking at prior art, specifications do not generally include these limitations as requirements.

> side note: the part you reference encourages to use ieee754 and mentions that anything else is not portable (= out of scope of the spec even if not forbidden), so guess we should do it which means using ieee754

No, this text defines no requirements. And most explicitly, the first sentence of that section makes my point.

```
This specification _allows implementations_ to set limits on the
range and precision of numbers accepted.
```

Implementations govern what they want to support.  The spec does not.

**Reply by rmannibucau** [reply]
> No. Having worked on multiple specifications, and looking at prior art, specifications do not generally include these limitations as requirements.

Well as a matter of fact a lot of spec do - all the one desiring to define a behavior.
Anyway, if it is not done it means numbers are almost not usable (floatting ones) for JSON-Logic and it is not testable (which is almost as not being defined so not supported by the spec).

> Implementations govern what they want to support. The spec does not.

This is true when the spec is not used only, the only role of the spec is to define explicitly what is supported with the consequence that if it is not in the spec it can be supported by an impl but it is not required so if you want to write something portable you must not use it.
An implementation can govern what it wants to support until it desire to claim to be compatible with the spec and there it must implement the spec requirements (or more but not less so if the spec states floatting numbers on 64 bits, a 32 bits impl will not be compatible with the spec).

So there is really no real way JSON-Logic relies on numbers without defining the scope it relies on - but agree it can open the door to impl to support more.

Side note: while I think IEEE754 is the right choice there this is not a strong technical constraint even if I guess most impl do it.

**Reply by gregsdennis** [reply]
JSON imposes no restrictions on numbers (which can be floating point), and it's obviously interoperable. [JSON Schema](https://json-schema.org/specification), [JSON Path](https://datatracker.ietf.org/doc/rfc9535/), and [JSON-e](https://json-e.js.org/) all use JSON's notion of numbers, imposing no further restrictions, and interoperability is core to their specs. All of these specs make good use of numbers as defined by JSON.

I'm not sure how you mean that numbers will be "almost unusable", that it's not testable, or that JSON Logic wouldn't be able to rely on numbers.  I see a lot of evidence to the contrary.

**Reply by JTeeuwissen** [reply]
I'm not familiar with json-e, but do any of these specs perform calculations with the json input numbers?
Since json logic does, and it feels like it's important for portability to know whether `{"+":[0.1,0.2]}` should result in `0.30000000000000004` or `0.3`

**Reply by rmannibucau** [reply]
Jsonschema does computations and....recommands to not use numbers but strings for that exact reason (https://github.com/json-schema-org/json-schema-spec/issues/312 for ex but google multipleof and decimal and jsonschema to get more examples).

So yes, the issue exists already everywhere and this is why the json spec doesnt put limitations but recommands to put one.

**Reply by gregsdennis** [reply]
> recommands to not use numbers but strings

No it does not.  I am the current primary author of the JSON Schema spec, and it most definitely does not make this recommendation.

The validation specification contains a paragraph briefly mentioning language support for numbers, and then only as an informational note in the context that such support should be to the extent that numbers in JSON are supported.

Secondly, none of the numeric keywords work on strings. They only work on numbers, _JSON_ numbers as defined by RFC8259.

**Reply by rmannibucau** [reply]
> Secondly, none of the numeric keywords work on strings. They only work on numbers, JSON numbers as defined by RFC8259.

Exactly! this is why you must define what must be supported by numbers.
Look all related issues it creates and implementations decide to not solve (here is just a random one https://github.com/ajv-validator/ajv/issues/652).

As a side note: there is no way to get an exact behavior _even using strings_ which is able to support way more than plain 64 bits "double" or alike, so if you don't put supported number definition there, there is no way to implement the spec cause of rounding or precision issues.
You can indeed write that in a spec but it should be highlighted and the related part must be associated with a warning "don't use it since the behavior is not the one we define in the spec".

This is a not pitfall of computers and there is no solution for unlimited precision (literally none even on supercalculators).

I'm fine just saying "take care the support of this feature can be limited by the precision of your runtime" but there can't be a "MUST" support X without more details on number - this is just technically impossible.
And if there is no "MUST" it is not in the spec and if it is not in the spec it doesn't need to be defined so ultimately without a number definition we don't need numbers at all.

**Reply by TotalTechGeek** [reply]
I think @JTeeuwissen had a pretty decent point here w/r/t `{"+":[0.1,0.2]}` and portability.

However, I think as I've had more time to process this; I've come to side even more with a purely mathematical definition of JSON Logic. So my `+1` is now a `++1`

Because we want JSON Logic to be useful as a general AST, I believe we should define both the operators and the data model in such a way where the implementation is responsible for handling (and potentially configuring) the mathematical precision.

I do believe that practically speaking, IEEE754 will be the default precision for most implementations; I think it would be ideal to list the supported precisions for implementations in our compatibility tables.

But I think cross-language adopters who are concerned about interoperability for numbers would like a setup where they can communicate: "Hey; we're using JSON Logic Core w/ IEEE754 Precision"

-- 

I can also foresee use-cases where IEEE754 is deemed unacceptable by an organization, and they might prefer to redefine the precision to use BigDecimal or something similar for their rules and calculations.

**Reply by rmannibucau** [reply]
> the implementation is responsible for handling (and potentially configuring) the mathematical precision

well, if you mean "making it work" it is just not possible and if you mean "putting its own constraints" it means the implementation must do a spec of the spec, I don't see it going very far if it is the path it takes so hope we can make something more useful for end users.

> I can also foresee use-cases where IEEE754 is deemed unacceptable by an organization, and they might prefer to redefine the precision to use BigDecimal or something similar for their rules and calculations.

Fully agree...but will mean in 80% of the case also using specific operators (if BigDecimal refers to java case, there is all this config about the rounding mode, precision etc to setup to make it functional) so this is already out of the spec and impl specific - and trust me it will end up being numbers in strings and not strings directly parsed as numbers just to make the json more interoperable (not a guess if ambiguous ;)).

**Reply by TotalTechGeek** [reply]
> and trust me it will end up being numbers in strings and not strings directly parsed as numbers just to make the json more interoperable

😆 

I have both a philosphical and practical view on this, that fortunately aren't in total conflict in this case.

Philosphically, I side with the article that @gregsdennis shared & how we should expect more from parsers, and we SHOULD avoid anti-patterns. 

Practically, systems aren't greenfield; I know there's going to be a mess of dependencies and multiple systems, _but_ fortunately JSON Logic's operators are defined with implicit coercion, 
> https://github.com/json-logic/.github/blob/0ee41ec8907f3fed9922908ade97f99974388c2e/tests/base.json#L514

Which means that if folks are worried about lossy truncation as data flows through different systems and services (w/ some codebases belonging to some team in some other country, with other priorities than fixing the JSON Parser), it'll still end up working out even if they embed the numbers as strings.

If you're working around limitations, rather than fixing them, you get weirdness 😬 . But I'd like to avoid holding other use-cases and implementations back.

> it means the implementation must do a spec of the spec

I see it as `Core<IEEE754>` in generic / templating terms. 

I think IEEE754 will practically be a widely supported baseline. From a org perspective, I think we should aim to have widely supported baselines. 

From a "Core" perspective & targeting general AST, I'd prefer to make precision an implementation detail / configuration.

**Reply by rmannibucau** [reply]
Well, parser can not help at all even assuming they can parse infinite numbers as string...issue is computers cant do math operations even with big numbers, without a precision loss at some point.
So if you enable math operations then you must define what will be the result - this is the minimum for operators.

Ill let you sleep on it but if the impl does the definition it is the spec, not an impl by definition ;).

There is also no holding putting limitations there, it defines the minimum expectation, impl can always enable more if desired using their config mecanism so from my window there are only drawbacks not specifying what will be the default expected behavior.

As an user i had use cases for ieee754 (precision was not a crazy requirement but perf was) and big decimal (account amount management).
My impl was backed by jsonp so was supporting this parser thing but as an user I still needed to pick one or the other and know upfront the behavior.
While bound to an impl, all good, but when you expose jsonlogic as an extension point it is very bad cause you just know it is jsonlogic which means nothing if undefined.
Ineeded vendor can write the spec but then what is the point of having a spec if it is not usable? I really fail to see how it help to say "we do thing but you will not know what" (exagering a bit but at the end it is what will happen when jsonlogic ends in extension points/config/callbacks).

Now a good point is: where do you put this constraint, it is either the model but I would also be happy if it ends in operator config - a bit like format of jsonschema or siblings...but it must be in the spec whereever it is.

**Reply by codetiger** [reply]
Am inclined towards "The spec defines the minimum expectation" and the  "impl can always enable more if desired" as mentioned by @rmannibucau. 

As someone using JSONLogic for financial message processing, I expect the spec to prescribe how default implementation in any language should work(e.g., by specifying IEEE754 as the baseline). And If the precision is not enough, spec should allow extending the implementation (like BigDecimal) using configurations. If the spec does not say the minimum expectation, we might end up having very different variations and user has to dig into the code to see what each implementation supports by default. For ex: Some implementations might support datetime + n by default while others don't, despite this being a common use case.

JSONLogic's focused scope as an AST spec actually makes it more suitable for such prescription compared to JSONSchema and other specs mentioned. While those specs need to accommodate a wider range of use cases, JSONLogic's more specific purpose allows us to be more prescriptive.

**Reply by TotalTechGeek** [reply]
> [!IMPORTANT]  
> This is being attached to the proposal.

I think something that I can improve on is helping to clarify some of the outputs of this initiative, and how we can accommodate these different end goals.

I think we're all agreed that:
- We'd probably like to support multiple precisions.
- We'd like to ensure portability.
- We'd like to build a robust Core Spec.
- We'd like to make it easy for folks to adopt JSON Logic, and know what they have available to them. 

I'm going to over-simplify to keep things short,

I believe that @gregsdennis's main desire in this project is to help build a decent and robust Core Spec / RFC. Specifications are often general to make it easy to adopt & adapt for different architectures and runtimes. And I really strongly back this.

It would be awesome if we could develop a simple, general purpose AST specification that works in many environments. (_Talking out my rear for a bit, I believe this would have useful applications in academia to introduce students to designing compilers & learning about syntax trees. But let's put this aside for a bit_)

The main concern I believe is being communicated is: 
- Without specifying a precision in the Core Spec, we could end up with poor compatibility between implementations.

Here's how I think we can address that:
- "JSON Logic Core" focuses on providing a specification of behaviors, from a pure mathematical perspective (numbers have infinite precision, etc). 
- The test cases designed for developers to test Core will have a multitude of math tests, but most will be integer-based and simple. The purpose of the tests will be to ensure proper application of the logic, and not specifically the precision employed by the implementation.
- This allows "JSON Logic Core" to remain portable and somewhat un-opinionated.

_(Now, before I continue, remember for a moment that "Core" will not include some deprecated operators that will be moved to a `Legacy` Community Extension.)_
- The JSON Logic Organization can mandate a "Recommended Baseline" for implementations, that can be updated as Community Extensions become more common. This is not part of the Core Specification or attached to an RFC.
- The JSON Logic Organization _could_ mandate that the "Recommended Baseline" include Core, the Legacy Extension, and support for IEEE754 Precision. 

This allows us to show in our compatibility tables which implementations meet our Recommended Baseline, and ideally promote implementations that also support alternative precisions.

**Reply by TotalTechGeek** [reply]
I'm appending the above suggestion to this proposal, so we can refer to it going forward. 

I'm going to revote with a +1.

**Reply by rmannibucau** [reply]
How would we make core profile usable by adding precisions (defined like that core is unusable by any app). Is the plan a number appendix for implementors? I dont think making it in another profile would work well.
Alternative is to not put math operators there and move that concern to "math extensions" meaning numbers are kept "absolute" and tests are almost only identity/val ones in core.

**Reply by dslmeinte** [reply]
I like this idea of putting all math operations in a `math` extension, or even an IEEE754-specific extension (`ieee754-arithmetic`). After all, IEEE754 are roughly as problematic as dates(+times) are.

**Reply by TotalTechGeek** [reply]
`-1` to moving it all to a math extension. I'm **very firmly opposed** to removing the math operators from the core spec, with great intensity. That also doesn't even really resolve the issue of how numbers are handled in other operators. If we remove everything from core, I don't think we really produce a viable output. 

Plus, the arrangement discussed for Core was that we'd define replacement operators for it; not just remove.

Everything that exists in `tests.json` should be able to be performed in Core.

**Reply by TotalTechGeek** [reply]
> How would we make core profile usable by adding precisions (defined like that core is unusable by any app)

Core is a specification. It's general. Like an abstract class or interface. `Core<IEEE754>` is an instantiation of it. Core does not include the legacy extension, but the implementations we promote will.

As an org, we can maintain a recommended baseline and recommend it in our compatibility tables.

For Core to be viable, we do not need to embed tests like `{ "+": [0.1, 0.2] }` equals `0.30000000000000004`. We can embed tests like `{ ">": [5, 3] }` and `{ "+": [1, 2, 3] }`

If we opinionate Core and say that it uses IEEE754, then we essentially communicate "Anything else is really not JSON Logic," which is not what we're going for. 

@gregsdennis's implementation of JSON Logic uses `decimal` instead of `double`. 

For pretty much everything, our implementations get all the same answers and calculations, we just have some different precisions on some Decimal calculations.

Because his implementation gets `0.3` instead of `0.30000000000000004`, I do not think that invalidates his implementation as Core compliant. 

If we want to add these opinions, we can do it from an org level with a Recommended Baseline.

--

Similarly, if we explore "profiles", I don't think profiles are a piece of Core, but are instead an org concept. I think we can maintain profiles and test for them within the organization and our compat tables 😄.

**Reply by rmannibucau** [reply]
> Core is a specification. It's general. Like an abstract class or interface. 

This is true but it also implies it exacty defines the expectations of its contract (in/out, not impl details) otherwise it is an API nobody can call.

> we do not need to embed tests like { "+": [0.1, 0.2] } equals 0.30000000000000004

actually it is exactly the kind of test which will make more value of the test suite than others.

> If we opinionate Core and say that it uses IEEE754, then we essentially communicate "Anything else is really not JSON Logic," which is not what we're going for.

Anything else is not JSON Logic *core* and it is the exact intent, like an CDI Lite application is not a CDI application nor a jakartaEE application but this distinction is what makes sense for consumers.

> Similarly, if we explore "profiles", I don't think profiles are a piece of Core, but are instead an org concept.

Agree, the key there would be to be able to stack profile (ie if you use profile "plus" you can also use "core" otherwise you need compatibility matrix and the nightmare starts).

back to core and math operators, the solution can be as easy as supporting for all operators an additional "opts" key and define that by default the mathematical operators imply `opts:{precision: ieee754}` or alike and enable to change it in other specs or impl.

So `{ "+": [0.1, 0.2] }` is equivalent to `{ "+": [0.1, 0.2], "opts":{"precision": "ieee754"} }` and is `0.30000000000000004` but if you  if set `precision` to something else you use another computation mecanism. Key there is to not define an enum but an extensible value implementations can rely on.

Ideally we'd need this opts at the root of the JSON-Logic aggregate to have global defaults:

`{version:1,opts: { math: {prevision: ieee754}, ... }, jsonlogic: { ...} }` would be the structure of an explicit and complete json-logic config.

**Reply by TotalTechGeek** [reply]
> Anything else is not JSON Logic core and it is the exact intent

Yes, this is **exactly** what I don't want. "Hey finance apps, you're not actually able to use JSON Logic; everything you're doing is off-spec and non-compliant." 

> otherwise it is an API nobody can call.

This isn't quite true; people definitely use JSON even though different implementations opt for decimal vs IEEE754, and for our math, we can definitely provide test cases that are precision-agnostic that are mostly integer based.

> , the solution can be as easy as supporting for all operators an additional "opts" key and define that by default the mathematical operators imply
> `{ "+": [0.1, 0.2], "opts":{"precision": "ieee754"} }`

I don't think multi-key operators are a good idea for JSON Logic. 

It introduces a ton of ambiguity in what the operator is vs what is an additional argument. Some implementations of JSON do not preserve key order, (postgres with `jsonb` will silently rearrange fields in your objects).

Setting precision on an operator level like this is also going to create headaches and serious divergence; "My json parser deserialized it to decimal but the operator math precision is ieee754, what are the proper casting rules here?"

It also doesn't map as well to S-Expressions.

>  opts at the root of the JSON-Logic aggregate

I don't think that most use-cases will desire this. I'm okay with these settings being set on the parser. 

IMO, if we did introduce a spec extension, it would be a valid JSON Logic operator, so a separate parser doesn't need to be built for it.


```json
{ "spec": ["1.2", "iee754", { "+": [1,2] }] }
```

## And to reiterate

I'm not against the organization setting policies and organizing a recommended baseline, and promoting implementations that are compliant with this baseline.

Core is meant to be a portable specification that is easy to adopt and adapt. 

We really don't need to set ourselves up in such a way where we alienate use-cases and implementations from being "JSON Logic".

I truly believe everyone can get what they want here 😅.

**Reply by rmannibucau** [reply]
Ok so we need the precision - either numbers or operators - in core (you dont want finance app exclusion but we dont want perf app exclusion too and both must use diff impl not converging well).

Multi keys operators are not that bad if based on an convention (know js often uses "options" or "opts" as key for the config). There is no real drawback and single key values are cool but technically not critical at all and not what makes jsonlogic great IMHO (it is more the extensibility and overall composability).
That said if you care just ensure it is an array instead of an object and there is an order solving the parser issue. If first item is an object it is options, if next (first or second) is a string it is the operator then the params.

There will be no real headache I think since impl are trivially handling that by keeping the number buffer while evaluating the precision, due to the small size of this bufger this is very acceptable and supported by all impl - and anyway not user facing so not important IMHO.

> Core is meant to be a portable specification that is easy to adopt and adapt.

What does it mean if it doesnt define behaviors?
Just that it is not portable nor adoptable (take back your finance example, if using a compatible impl i dont know if im loosing precision can i use it? Clearly not. And moving this responsability to the impl means any product must expose its impl and never change it...i fail to see how it can relate to anything real life.

Ok for spec format.

**Reply by JTeeuwissen** [reply]
> Core is meant to be a portable specification that is easy to adopt and adapt.

Not requiring implementations to adhere to a number spec sounds fine for json logic the "concept", but I wouldn't call this portable. Perhaps there is a better naming scheme to differentiate between the two? I suggest
- Json logic core (conceptual, not portable, ast, syntax, simple rule overview, precision agnostic)
- Json logic standard (portable number spec, precision tests, perhaps string encoding requirements)

This way using decimals could still be called Json logic, but it could be made clear that it is not portable. This would also communicate a clear "works for most" feature set, for users and implementors.

**Reply by TotalTechGeek** [reply]
> What does it mean if it doesnt define behaviors?

But it would define behaviors! 😄  

> Convergence

Both my decimal and IEEE754 implementations both pass a couple hundred JSON Logic tests (and certainly everything in `tests.json`). 

We can define our math operators using pure math. `+` is addition, `-` is subtraction, so on and so forth. We can talk about how they're variadic, and define a robust set of test cases that are precision agnostic.

I don't think that for the purposes of Core, that we need to explicitly declare that `{ "+": [0.1, 0.2] }` is some quirky number. I think precision being configurable for the AST is actually a feature, and not a bug.

**Reply by TotalTechGeek** [reply]
> but it could be made clear that it is not portable. 

But why shouldn't it be? I'd contribute development resources to try to ensure cross implementation compatibility and opening JSON Logic up for different use-cases.

> Json logic core (ast, syntax, simple rule overview, precision agnostic)
> Json logic standard (number spec, precision tests, perhaps string encoding requirements)

There's already:
- AST (AST, Syntax)
- Core (Operators, Defined in a Precision Agnostic way)

And again, even if we rebrand standard and core, "standard" as proposed above still alienates use cases that require better precision, making adoption more uncomfortable.

The AST is barely useful on its own for an RFC. 

I'm okay with introducing:
- "Recommended Baseline"

For tracking implementations, or "Profiles"

**Reply by JTeeuwissen** [reply]
I wouldn't call it alienating users, but I think it makes sense to pick a default that works for the majority of users that should be supported by all json logic implementations. With optional configurability to change e.g. numbers to decimals. But users shouldn't have to know more than "adheres to spec" when using Json logic across multiple implementations.

**Reply by TotalTechGeek** [reply]
> But users shouldn't have to know more than "adheres to spec" when using Json logic across multiple implementations.

This is essentially what I'm promoting though. 

JSON Logic as an org would mandate its recommended baseline / main profile / whatever we want to call it, but the IETF proposal or whatever route we go with is generic enough to be architecture / runtime portable.

That way you get `AST -- Core -- Community 1.1`.

And the community profile can be promoted easily to include other common extensions that are supported later on, without revising core. (If we add ISO date extensions or something)

**Reply by gregsdennis** [reply]
For the most explicitly clarity on this issue, I present the `multipleOf` excerpt from the spec and a test case from the JSON Schema Test Suite.

> [6.2.1.](https://json-schema.org/draft/2020-12/json-schema-validation#section-6.2.1) multipleOf
> 
> The value of "multipleOf" MUST be a number, strictly greater than 0.
> A numeric instance is valid only if division by this keyword's value results in an integer.

```json
{
  "description": "by small number",
  "schema": {
      "$schema": "https://json-schema.org/draft/2020-12/schema",
      "multipleOf": 0.0001
  },
  "tests": [
      {
          "description": "0.0075 is multiple of 0.0001",
          "data": 0.0075,
          "valid": true
      },
      {
          "description": "0.00751 is not multiple of 0.0001",
          "data": 0.00751,
          "valid": false
      }
  ]
}
```

Just about every implementation of JSON Schema across a multitude of languages passes this test case, despite many of them using IEEE754 double-precision floats.

In C#, to illustrate the difference, I created this test:

```c#
double a = .0075;  // IEEE754
double b = .0001;
decimal c = .0075m;
decimal d = .0001m;

Console.WriteLine("double");
Console.WriteLine(a % b);
Console.WriteLine();
Console.WriteLine("decimal");
Console.WriteLine(c % d);
```

which outputs

```
double
9.999999999999937E-05

decimal
0.0000
```

You can see that simply using the built-in modulus operator `%` presents a problem with IEEE754 math (which is what @rmannibucau is arguing).  Despite this, the implementations pass these tests.

JSON Schema is not made unusable by requiring a mathematical interpretation of division for `multipleOf`.  I also argue that most people find this easier to understand.

Implementations are not burdened with any kind of difficulty in supporting this, and it's easy to understand because it's very simple math.

I grant that a "multipleOf" calculation isn't as extensive as the math that JSON Logic offers, but my point is the same: defining the spec based on math makes sense to most people.  Enforcing it with the test suite will ensure that implementations align.  It may put a small burden on the implementations, but that's where the burden should lie, not with the users.

**Reply by JTeeuwissen** [reply]
Just as a note, the c# implementation of json schema from ricosuter uses decimal for multipleOf.

**Reply by rmannibucau** [reply]
Just means the test suite coverage is bad and doesnt consider edge cases which lead to all issues (just check on github) so this is  abad practise for a spec.

**Reply by rmannibucau** [reply]
@JTeeuwissen moving to 16bytes just moves the issue, if spec says it is infinite as proposed your impl still has issues. Even in string based impl you have limits on the precision.

**Comment by TotalTechGeek** [comment]
> [!Warning]
> Revote necessary because of new appended suggestion; this is an outdated vote.

Approved; +1

**Comment by codetiger** [comment]
> [!Warning]
> Revote necessary because of new appended suggestion; this is an outdated vote.

Agreed 100%. 
One of the reasons why I named my repo DataLogic-rs is to support the model is different formats.

**Comment by TotalTechGeek** [comment]
Okay: We have 3 TC Approvals (I'm assuming @gregsdennis implicitly approves this), but some community dissent.

I think I'd like to switch to Apache Voting,

> [!Warning]
> Revote necessary because of new appended suggestion; these are outdated votes.

TC Approval Votes: 
- @gregsdennis (Assumed +1)
- @TotalTechGeek (+1)
- @codetiger (+1) 

Community Dissent:
- @rmannibucau (Assumed -1) 

Assuming @JTeeuwissen is a -0.2, and @dslmeinte is a -0.1 (maybe +0.1)

The dissent is on whether the JSON Logic Core specification should incorporate any additional specifications around "numbers" beyond what JSON specifies.


### Dissent Summary 

Community Concern: Because JSON does not specify a range for numbers, many JSON parsers use IEEE754, and cannot represent large integers.

@gregsdennis believes that the JSON / YAML Parser is responsible for deciding the range, not the JSON Logic Core spec, which follows prior art with what JSON / JSON-E / JSON Schema decided.

**Reply by TotalTechGeek** [reply]
Because we haven't "officially" tweaked our voting system, I'm going to use the previously communicated rules for "If there's dissent, we'll try to reach consensus, and if it can't be reached, a 4th TC member's vote is required"

**Reply by rmannibucau** [reply]
FTR i'm not against, I'm just pointing out that "numbers" doesn't mean anything if not defined

**Reply by TotalTechGeek** [reply]
Ah; apologies if I assumed `-1` a bit overzealously! 

I'm not sure if I find myself disagreeing with the community dissent myself, but I think I'm comfortable with the convention exercised by other projects, and I'd be for incorporating the supported ranges into our compatibility tables.

**Reply by toddbaert** [reply]
Consider me on-board as well.

**Comment by TotalTechGeek** [comment]
Okay folks! I don't think we're going to reach perfect consensus here.

I think everyone had **great** input, and I think debate truly helps us achieve a better outcome. 

To be clear, I did appreciate the points / thoughts raised by @JTeeuwissen, @dslmeinte and @rmannibucau, and I hope everyone felt heard in this discussion. 

I think based on the approvals, we will be moving forward with the compromise:
- Core remains precision agnostic and a general specification. 
- [JSON Logic](https://github.com/json-logic) will maintain profiles and a recommended standard for implementations.
- If we introduce a `spec` operator, named profiles will come from JSON Logic organization, not from "Core". 

The points about the logic & behavior portability are acknowledged; and we agree that there should be a mandate, but the implementation threshold will be maintained outside of the RFC. 

This will allow the org to upgrade the base threshold more rapidly as we get mass support for community extensions.

--- 

TC Approvers:
- @TotalTechGeek (+1)
- @toddbaert (+1
- @gregsdennis (+1)

Assumed Abstaining:
- @codetiger

**Reply by TotalTechGeek** [reply]
imo; this was one of four major proposals we need to knock out before we begin the mass re-evaluation of operators, though another member might spot something I haven't.

We needed to decide:
- Backwards Compat Policy ✅ 
- Data model / precision policies for JSON Logic ✅ 
- Do we replace `var` with something else?
_(To define tests properly, getting out context operators locked in is relatively critical)_
- Do we erase sugaring from JSON Logic? (This seems to be a regret of @jwadhams)
_(https://github.com/jwadhams/json-logic/issues/44#issuecomment-2082929035. My prescribed approach is rather different from Jeremy's, but I do think a change should be made. In compliance with the policy, the change should also be backwards compatible)_

Once we decide these guys, I think it opens the flood gates for re-evaluating each operator and designing a robust suite of tests for each of them.

**Reply by codetiger** [reply]
After the discussion I am changing my vote to +1 for the proposal.
